src/main/java/org/spongepowered/api/Tag.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api;
+
+import com.google.common.reflect.TypeToken;
+import org.spongepowered.api.block.BlockTags;
+
+import java.util.Collection;
+
+/**
+ * Represents a tag which can be applied to a specific {@link CatalogType}.
+ *
+ * @param <C> The catalog type
+ * @see BlockTags
+ */
+public interface Tag<C extends CatalogType> extends CatalogType {
```

Comment:
Can't enforce the `Tag` to only contain `CatalogType`s. Vanilla doesn't have a unified interface for `Tag` to use, so it's just `Tag<T>` sadly. Tags however can be catalog typed.

Reply:
If tags are not bound to a single type, as indicated by your next comment then you could simply skip the type parameter.

Reply:
Vanilla tag entries are all namespaced afaik, so they do actually contain catalog types.

Reply:
> Vanilla tag entries are all namespaced afaik

But there's no unique interface to attach to them... I could literally make a `Tag<Entity>` and the game will be happy with that tag, even though it's not a valid use of the tag. Similarly, Sponge cannot enforce that the returned types are `CatalogType`s as mods can introduce new Tags of their own classes, which would immediately fail class casts to `CatalogType`.

Reply:
Easy to work around that with the subtypes and wrappers. We don't have to do everything 1:1 with how things are done in vanilla mc.

Reply:
> Easy to work around that with the subtypes and wrappers.

Yes, it's easy to work around with, but it's extra maintenance and potentially bug inducing that I don't want to have to deal with. Since mods already do all sorts of things by having near unlimited access to internals, compared to plugins interacting with the API, I don't want there to be any room for overhead of maintenance costs for a wrapper system, let alone subtypes for an API that is supposed to represent/expose functionality implemented by vanilla. An easy example of how subtyping ends up adding extreme overhead of maintenance costs is the MaterialData/ItemMeta API in Bukkit, and I can tell you right now that that system was already an issue to represent mod provided datas.



> We don't have to do everything 1:1 with how things are done in vanilla mc.

You're right, we're not doing everything 1:1, but core functionality does dictate a lot in how our API can be designed, and the direction in which it is being designed. We've taken a lot of queues in our original API design to avoid wrapping for most things, and lately with 1.13's world gen changes, @Zidane and I already came to the agreement that we'd be making it a point to avoid wrapping at most all costs. These days, looking at the implementation, the biggest offenders of wrapping basically anything is to support API designs that are not representative of their implementation counterparts (Inventory and Data). This bit of API however, is **supposed** to represent the Tag system from vanilla, much like the system that was introduced for what was previously known as [`BlockTrait`s](https://github.com/SpongePowered/SpongeAPI/commit/1d4d3cff300003b9353a0d0415b8345e904da5b2#diff-4e4e8fb34984746e53f060d9a7a78eeb) because it needed to be representable for what is universally used in the implementation (and happily coexists with Data API's counterparts with keys and values).

Reply:
To add on, other examples of wrapper implemented APIs that have since disappeared: Text API implementation (used to be fully implemented in the API, now it's all interfaced onto the implementation using TextComponents as they should), Populators (quite literally re-implemented all vanilla populators with a mix and match of some implementation provided mixed in classes, and others that were full re-implementations using sponge classes), SpawnCauses (eliminated with the context system), etc.

Reply:
Without subtypes, we can't make a Tag a catalogue type because we end up mixing different registries. We could add a prefix like `blocks/<id>` to resolve this, but then we'll still need to get it from somewhere, which can the cause the same problems with mods. This also mismatches ids with vanilla mc. IMO these are different catalog types and shouldn't be mixed, they have different registries in vanilla mc as well. There will always be a workaround to get this done properly if you want them to be catalog types, subtypes or not.

@kashike Maybe you can share your thoughts?

Reply:
> Without subtypes, we can't make a Tag a catalogue type because we end up mixing different registries.

Looking further, it's impossible then to provide all Tags regardless whether they're catalogued for Blocks, Items, or Fluids, since Mods will be able to create their own tag collections and using their own tags, being able to expose tags that are not registered in any of the three default "registries".



> We could add a prefix like `blocks/<id>` to resolve this, but then we'll still need to get it from somewhere, which can the cause the same problems with mods.

Right, because even a mod could make their own TagCollection of Tag<Item> and synchronize it on their own parts, which ends up with sponge not knowing how to represent that tag. I'd say it's easiest then to just have `Tag` exposing it's `CatalogKey` (which it always has to have), and that's about it. We can expose the `TagCollection` object and expose the three types of vanilla ones, but they are all temporarily unable to be `CatalogType`d to be searchable via `GameRegistry`. Given the dynamic data driven nature of Tags, I'd say this is the only way to go for now.

Reply:
If we expose `TagCollection`, we end up with yet another different system. Which is inconsistent with how the rest of the API works. And they could be perfectly catalogued for the ones we know about. We shouldn't care about the custom tag collections that are created by mods.